`pnpm add --allow-build` replaces existing `allowBuilds` config in pnpm-workspace.yaml

### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

_No response_

### Which area(s) of pnpm are affected? (leave empty if unsure)

_No response_

### Link to the code that reproduces this issue or a replay of the bug

_No response_

### Reproduction steps

```bash
$ mkdir z

$ cd z

$ pnpm add unrs-resolver --allow-build=unrs-resolver

$ cat pnpm-workspace.yaml
allowBuilds:
  unrs-resolver: true

$ pnpm add esbuild --allow-build=esbuild

$ cat pnpm-workspace.yaml
allowBuilds:
  esbuild: true

$ rm -rf node_modules

$ pnpm install
✓ Lockfile passes supply-chain policies (verified 26s ago)
Lockfile is up to date, resolution step is skipped
Packages: +5
+++++
Packages are cloned from the content-addressable store to the virtual store.
  Content-addressable store is at: /Users/k/Library/pnpm/store/v11
  Virtual store is at:             node_modules/.pnpm
Progress: resolved 5, reused 5, downloaded 0, added 5, done
node_modules/.pnpm/esbuild@0.28.1/node_modules/esbuild: Running postinstall script, done in 755ms

dependencies:
+ esbuild 0.28.1
+ unrs-resolver 1.12.2

[ERR_PNPM_IGNORED_BUILDS] Ignored build scripts: unrs-resolver@1.12.2

Run "pnpm approve-builds" to pick which dependencies should be allowed to run scripts.
```

### Describe the Bug

The `--allow-build` flag for `pnpm add` replaces the `allowBuilds` config in `pnpm-workspace.yaml` instead of adding to it / modifying it

This is a regression - we were relying on the previous behavior

### Expected Behavior

The `--allow-build` flag for `pnpm add` modifies the existing `allowBuilds` config in `pnpm-workspace.yaml`

### Which Node.js version are you using?

v24.13.0

### Which operating systems have you used?

- [x] macOS
- [ ] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

--- a/.changeset/add-allow-build-keeps-existing-entries.md
+++ b/.changeset/add-allow-build-keeps-existing-entries.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/installing.commands": patch
+"pnpm": patch
+---
+
+`pnpm add --allow-build` now adds to the `allowBuilds` entries already in `pnpm-workspace.yaml` instead of replacing them [#13872](https://github.com/pnpm/pnpm/issues/13872).

--- a/pnpm11/installing/commands/src/add.ts
+++ b/pnpm11/installing/commands/src/add.ts
@@ -293,7 +293,7 @@ export async function handler (
         })
       }
     }
-    const allowBuilds: Record<string, boolean> = {}
+    const allowBuilds = { ...opts.allowBuilds }
     for (const pkg of opts.allowBuild) {
       allowBuilds[pkg] = true
     }
@@ -307,14 +307,9 @@ export async function handler (
         },
       })
     }
-    // Pass the allowed packages to allowBuilds so they can build during this install
-    const mergedAllowBuilds = { ...opts.allowBuilds }
-    for (const pkg of opts.allowBuild) {
-      mergedAllowBuilds[pkg] = true
-    }
     await installDeps({
       ...opts,
-      allowBuilds: mergedAllowBuilds,
+      allowBuilds,
       rebuildHandler: commands?.rebuild,
       include,
       includeDirect: include,

--- a/pnpm11/pnpm/test/install/lifecycleScripts.ts
+++ b/pnpm11/pnpm/test/install/lifecycleScripts.ts
@@ -188,7 +188,7 @@ test('--allow-build flag keeps the packages already listed in allowBuilds', asyn
   writeYamlFileSync('pnpm-workspace.yaml', {
     allowBuilds: { '@pnpm.e2e/install-script-example': true },
   })
-  execPnpmSync(['add', '--allow-build=@pnpm.e2e/pre-and-postinstall-scripts-example', '@pnpm.e2e/pre-and-postinstall-scripts-example@1.0.0'])
+  execPnpmSync(['add', '--allow-build=@pnpm.e2e/pre-and-postinstall-scripts-example', '@pnpm.e2e/pre-and-postinstall-scripts-example@1.0.0'], { expectSuccess: true })
 
   const modulesManifest = await readWorkspaceManifest(project.dir())
   expect(modulesManifest?.allowBuilds).toStrictEqual({

--- a/pnpm11/pnpm/test/install/lifecycleScripts.ts
+++ b/pnpm11/pnpm/test/install/lifecycleScripts.ts
@@ -190,8 +190,8 @@ test('--allow-build flag keeps the packages already listed in allowBuilds', asyn
   })
   execPnpmSync(['add', '--allow-build=@pnpm.e2e/pre-and-postinstall-scripts-example', '@pnpm.e2e/pre-and-postinstall-scripts-example@1.0.0'], { expectSuccess: true })
 
-  const modulesManifest = await readWorkspaceManifest(project.dir())
-  expect(modulesManifest?.allowBuilds).toStrictEqual({
+  const workspaceManifest = await readWorkspaceManifest(project.dir())
+  expect(workspaceManifest?.allowBuilds).toStrictEqual({
     '@pnpm.e2e/install-script-example': true,
     '@pnpm.e2e/pre-and-postinstall-scripts-example': true,
   })

--- a/pnpm11/pnpm/test/install/lifecycleScripts.ts
+++ b/pnpm11/pnpm/test/install/lifecycleScripts.ts
@@ -183,6 +183,20 @@ test('selectively allow scripts in some dependencies by --allow-build flag', asy
   })
 })
 
+test('--allow-build flag keeps the packages already listed in allowBuilds', async () => {
+  const project = prepare({})
+  writeYamlFileSync('pnpm-workspace.yaml', {
+    allowBuilds: { '@pnpm.e2e/install-script-example': true },
+  })
+  execPnpmSync(['add', '--allow-build=@pnpm.e2e/pre-and-postinstall-scripts-example', '@pnpm.e2e/pre-and-postinstall-scripts-example@1.0.0'])
+
+  const modulesManifest = await readWorkspaceManifest(project.dir())
+  expect(modulesManifest?.allowBuilds).toStrictEqual({
+    '@pnpm.e2e/install-script-example': true,
+    '@pnpm.e2e/pre-and-postinstall-scripts-example': true,
+  })
+})
+
 test('--allow-build flag should specify the package', async () => {
   const project = prepare({})
   const result = execPnpmSync(['add', '@pnpm.e2e/pre-and-postinstall-scripts-example@1.0.0', '--allow-build'])

